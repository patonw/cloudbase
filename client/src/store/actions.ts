import { Action } from "redux"
import { UUID, Cell } from './types'

export enum AsyncStatus {
  Cached,
  Pending,
  Success,
  Failure,
}

export const CLEAR_ERROR = 'CLEAR_ERROR'
export const START_APP = 'START_APP'
export const CREATE_WORKSHEET = 'CREATE_WORKSHEET'
export const CREATE_PROCESS = 'CREATE_PROCESS'
export const LOAD_TOC = 'LOAD_TOC'
export const LOAD_WORKSHEET = 'LOAD_WORKSHEET'
export const EXECUTE_CELL = 'EXECUTE_CELL'
export const CREATE_CELL = 'CREATE_CELL'
export const DIRTY_CELL = 'DIRTY_CELL'
export const DIRTY_GRAPH = 'DIRTY_GRAPH'
export const COMMIT_CELL = 'COMMIT_CELL'
export const AFTER_COMMIT = 'AFTER_COMMIT'
export const INSERT_CODE_CELL = 'INSERT_CODE_CELL'
export const INSERT_GRAPH_CELL = 'INSERT_GRAPH_CELL'
export const DELETE_CELL = 'DELETE_CELL'
export const REORDER_WORKSHEET = 'REORDER_WORKSHEET'

export const LOCK_WORKSHEET = 'LOCK_WORKSHEET'
export const UNLOCK_WORKSHEET = 'UNLOCK_WORKSHEET'

// TODO Use generics for concrete actions
export interface ActionOf<T> extends Action {
  data: T
}

export interface AsyncActionOf<T> extends ActionOf<T> {
  status: AsyncStatus
  error?: any
}

export interface LoadTOCActionData {
  workbooks: {
    uuid: UUID
    name: string
    sheets: UUID[]
  }[]
  sheets: {
    uuid: UUID
    name: string
    workbook: UUID
  }[]
  processes: {
    [key: string]: UUID
  }
}

export interface AsyncAction extends Action {
  type: string
  status: AsyncStatus
  error?: any
}

export const isAsyncAction = (action: Action): action is AsyncAction => (action as AsyncAction).status !== undefined

export interface LoadTOCAction extends AsyncAction {
  type: typeof LOAD_TOC
  status: AsyncStatus
  data?: LoadTOCActionData
  error?: any
}

export type CreateWorksheetAction = AsyncActionOf<{
  bookId: UUID,
  name: string
}>

export interface LoadWorksheetActionData {
  uuid: UUID
  name: string
  workbook: UUID
  cells: {
    __typename: string
    uuid: UUID
    script: string
    spec?: string
  }[]
}

export interface LoadWorksheetAction extends AsyncAction {
  type: string
  uuid: UUID
  status: AsyncStatus
  data?: LoadWorksheetActionData
  error?: any
}

export interface CreateCellAction extends AsyncAction {
  type: string
  status: AsyncStatus
  worksheet: UUID
  cells?: UUID[] // New cell ordering
  data?: Cell
}

export interface ExecuteCellAction extends AsyncAction {
  type: string
  status: AsyncStatus
  uuid: UUID
  pid: UUID
  data?: string
  json?: () => any
  error?: any
}

export interface DirtyCellAction extends Action {
  type: string
  uuid: UUID
  script: string
}

export interface DirtyGraphAction extends Action {
  type: string
  uuid: UUID
  spec: string
}

export interface CommitCellAction extends Action {
  type: string
  status: AsyncStatus
  uuid: UUID
}

export type CreateProcessAction = AsyncActionOf<{
  sheetId: UUID,
  procId?: UUID,
}>

export type InsertGraphCellAction = AsyncActionOf<{
  sheetId: UUID
  index?: number
}>

export type InsertCodeCellAction = AsyncActionOf<{
  sheetId: UUID
  index?: number
  script?: string
}>


export type InsertCellResult = AsyncActionOf<{
  sheetId: UUID
  cells: UUID[]
  cell: Cell
}>

export type DeleteCellAction = AsyncActionOf<{
  sheetId: UUID
  cellId: UUID
}>

export type AfterCommitAction = ActionOf<Cell>

export type ReorderWorksheetAction = AsyncActionOf<{
  sheetId: UUID
  cells: UUID[]
}>

export function later(delay: number, value: any = null) {
  return new Promise(resolve => setTimeout(resolve, delay, value));
}

export function clearError(): Action {
  return {
    type: CLEAR_ERROR,
  }
}

export const startApp = () => ({ type: START_APP })

export const loadToc = (status = AsyncStatus.Pending) => ({
  type: LOAD_TOC,
  status,
})

export const createWorksheet = (bookId: UUID, name: string, status = AsyncStatus.Pending) => ({
  type: CREATE_WORKSHEET,
  status,
  data: {
    bookId,
    name,
  }
})

export const loadWorksheet = (uuid: UUID, status = AsyncStatus.Pending) => ({
  type: LOAD_WORKSHEET,
  status,
  uuid,
})

export const executeCell = (processId: UUID, cellId: UUID, status = AsyncStatus.Pending) => ({
  type: EXECUTE_CELL,
  uuid: cellId,
  pid: processId,
  status,
})

export const dirtyCell = (uuid: UUID, script: string) => ({
  type: DIRTY_CELL,
  uuid,
  script
})

export const dirtyGraph = (uuid: UUID, spec: string) => ({
  type: DIRTY_GRAPH,
  uuid,
  spec,
})

export const commitCell = (uuid: UUID, status = AsyncStatus.Pending) => ({
  type: COMMIT_CELL,
  uuid,
  status
})

export const afterCommit = (data: Cell) => ({
  type: AFTER_COMMIT,
  data,
})

export function insertCodeCell(sheetId: UUID, index?: number, status = AsyncStatus.Pending): InsertCodeCellAction {
  return {
    type: INSERT_CODE_CELL,
    status,
    data: {
      sheetId,
      index,
    }
  }
}

export function insertGraphCell(sheetId: UUID, index?: number, status = AsyncStatus.Pending): InsertGraphCellAction {
  return {
    type: INSERT_GRAPH_CELL,
    status,
    data: {
      sheetId,
      index,
    }
  }
}
export function deleteCell(sheetId: UUID, cellId: UUID, status = AsyncStatus.Pending): DeleteCellAction {
  return {
    type: DELETE_CELL,
    status,
    data: {
      sheetId,
      cellId,
    }
  }
}

export function reorderWorksheet(sheetId: UUID, cells: UUID[], status = AsyncStatus.Pending): ReorderWorksheetAction {
  return {
    type: REORDER_WORKSHEET,
    status,
    data: {
      sheetId,
      cells,
    }
  }

}

export function createProcess(sheetId: UUID, procId?: UUID,status = AsyncStatus.Pending) {
  return {
    type: CREATE_PROCESS,
    status,
    data: {
      sheetId,
      procId,
    }
  }
}

export function lockWorksheet() {
  return {
    type: LOCK_WORKSHEET
  }
}

export function unlockWorksheet() {
  return {
    type: UNLOCK_WORKSHEET
  }
}