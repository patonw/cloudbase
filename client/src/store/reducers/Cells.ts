import { Action } from "redux"

import { Cell } from '../types'

interface CellsState {
  [key: string]: Cell
}

const initialCellsState = {
  "45bn892vn90": {
    uuid: "45bn892vn90",
    worksheet: "w049tu34589",
    lang: "kotlin",
    script: "listOf(1,2,3)",
    result: null,
  },
  "endnut56nnd56u": {
    uuid: "endnut56nnd56u",
    worksheet: "w049tu34589",
    lang: "kotlin",
    script: "listOf(1,8,3)",
    result: null,
  },
  "r6rm68r76m8r76m8": {
    uuid: "r6rm68r76m8r76m8",
    worksheet: "w049tu34589",
    lang: "kotlin",
    script: "data class Foobar(val x)",
    result: null,
  },
  "q34v4qb4q3456w467n": {
    uuid: "q34v4qb4q3456w467n",
    worksheet: "w049tu34589",
    lang: "kotlin",
    script: "listOf(1,2,3)",
    result: null,
  },
}

export default function cellsReducer(state: CellsState = initialCellsState, action: Action) {
  return state
}