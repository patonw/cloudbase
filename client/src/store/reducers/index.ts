import { combineReducers } from "redux"

import cellsReducer from './Cells'
import viewStateReducer from './ViewState'
import workbooksReducer from './Workbooks'
import worksheetsReducer from './Worksheets'
import resultsReducer from "./Results";
import dirtyReducer from "./Dirty";

export const rootReducer = combineReducers({
  view: viewStateReducer,
  workbooks: workbooksReducer,
  worksheets: worksheetsReducer,
  cells: cellsReducer,
  results: resultsReducer,
  dirty: dirtyReducer,
})

export type AppState = ReturnType<typeof rootReducer>