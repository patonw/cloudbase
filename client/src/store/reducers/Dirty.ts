import { Action } from "redux"
import produce from "immer"

import { Cell } from '../types'
import { DIRTY_CELL, DirtyCellAction,  DIRTY_GRAPH, DirtyGraphAction, AFTER_COMMIT, AfterCommitAction } from "../actions";

interface DirtyState {
  [key: string]: Cell
}

export default function dirtyReducer(state: DirtyState = {}, action: Action) {
  return produce(state, draft => {
    switch (action.type) {
      case DIRTY_CELL: {
        const act = action as DirtyCellAction
        const { uuid, script } = act

        draft[uuid] = {
          ...draft[uuid],
          uuid,
          script,
        }

        break
      }

      case DIRTY_GRAPH: {
        const act = action as DirtyGraphAction
        const { uuid, spec } = act

        draft[uuid] = {
          ...draft[uuid],
          uuid,
          spec,
        }

        break
      }

      case AFTER_COMMIT: {
        const act = action as AfterCommitAction
          const { uuid } = act.data
          delete draft[uuid]
        break
      }
    }
  })
}