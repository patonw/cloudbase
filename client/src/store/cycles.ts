import { Action } from 'redux';
import { LOAD_WORKSHEET, AsyncAction, AsyncStatus, LoadWorksheetAction, EXECUTE_CELL, ExecuteCellAction, afterCommit, DELETE_CELL, DeleteCellAction, deleteCell, INSERT_CODE_CELL, InsertCodeCellAction, InsertCellResult, INSERT_GRAPH_CELL, InsertGraphCellAction, ReorderWorksheetAction, REORDER_WORKSHEET, reorderWorksheet } from './actions';

import * as gql from './graphql'
import { combineCycles } from 'redux-cycles';
import * as fp from 'lodash/fp'
import { Stream } from 'most'
import * as most from 'most'
import { AppState } from './reducers';
import { RequestInput } from '@cycle/http';

function ofType(tag: any) {
  return (action: Action) => action.type === tag
}

function ofStatus(tag: AsyncStatus) {
  return (action: AsyncAction) => action.status === tag
}

interface ResponseSuccess {
  body: any
  request: any
}

interface ResponseError {
  response: ResponseSuccess
}

type Response = ResponseSuccess | ResponseError
const isOK = (resp: Response): resp is ResponseSuccess => !fp.isError(resp)

function flatCatch(resp$$: Stream<Stream<Response>>): Stream<Response> {
  return resp$$
    .flatMap((resp$) =>
      resp$
        .recoverWith((err: ResponseError) => most.of(err))
    )
}

export function loadWorksheetCycle(sources: any) {
  const request$: Stream<RequestInput> = sources.ACTION
    .filter(ofType(LOAD_WORKSHEET))
    .filter(ofStatus(AsyncStatus.Pending))
    .map((action: LoadWorksheetAction) => ({
      url: '/graphql',
      category: LOAD_WORKSHEET,
      method: "POST",
      send: gql.loadWorksheet(action.uuid),
      uuid: action.uuid,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.worksheet", resp)
    }))

  const result$ = (sources.HTTP)
    .select(LOAD_WORKSHEET)
    .flatMap((response$: Stream<Response>) =>
      response$
        .recoverWith((err: ResponseError) => most.of(err))
    )
    .map((resp: Response) => {
      let uuid = fp.get("body.data.worksheet.uuid", resp)

      if (isOK(resp) && !!uuid) {
        return {
          type: LOAD_WORKSHEET,
          status: AsyncStatus.Success,
          uuid,
          data: fp.get("body.data.worksheet", resp)
        }
      }

      uuid = uuid || fp.get("response.request.uuid", resp)
      let error = fp.get("body.errors", resp) || fp.get("response.body.errors", resp) || resp
      return {
        type: LOAD_WORKSHEET,
        status: AsyncStatus.Failure,
        uuid: uuid,
        error,
      }
    })

  return {
    HTTP: request$,
    ACTION: result$
  }
}

function afterCommitCycle(sources: any) {
  const state$: Stream<AppState> = sources.STATE
  const action$: Stream<Action> = sources.ACTION

  const dirtyCells$ = state$.map(state => state.dirty)
  const execSuccess$ = action$
    .filter(ofType(EXECUTE_CELL))
    .map(it => it as ExecuteCellAction)
    .filter(ofStatus(AsyncStatus.Success))

  const resolved$ = execSuccess$
    .sample(
      (action, dirty) => dirty[action.uuid],
      execSuccess$,
      dirtyCells$
    )
    .filter(fp.identity)
    .map(dirtyCell => afterCommit(dirtyCell))

  return {
    ACTION: resolved$
  }
}

function deleteCellCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const request$: Stream<RequestInput> = action$
    .filter(ofType(DELETE_CELL))
    .map(act => act as DeleteCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({sheetId, cellId}) => gql.deleteCell(sheetId, cellId))
    .map(body => ({
      url: '/graphql',
      category: DELETE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.deleteCell.cell.uuid", resp)
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(DELETE_CELL)
    .thru(flatCatch)

  const reaction$ = resp$
    .flatMap(resp => {
      const cellId: string | undefined = fp.get("body.data.deleteCell.cell.uuid", resp)
      const sheetId: string | undefined = fp.get("body.data.deleteCell.sheet.uuid", resp)

      // TODO handle errors

      if (cellId && sheetId)
        return most.of(deleteCell(sheetId, cellId, AsyncStatus.Success))

      return most.empty()
    })

  return {
    HTTP: request$,
    ACTION: reaction$,
  }
}

function insertCellCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const codeReq$: Stream<RequestInput> = action$
    .filter(ofType(INSERT_CODE_CELL))
    .map(act => act as InsertCodeCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({sheetId, index}) => gql.insertCodeCell(sheetId, index))
    .map(body => ({
      url: '/graphql',
      category: INSERT_CODE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.insertCell.cell.uuid", resp)
    }))

  const graphReq$: Stream<RequestInput> = action$
    .filter(ofType(INSERT_GRAPH_CELL))
    .map(act => act as InsertGraphCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({sheetId, index}) => gql.insertGraphCell(sheetId, index))
    .map(body => ({
      url: '/graphql',
      category: INSERT_CODE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.insertCell.cell.uuid", resp)
    }))

  const request$ = most.merge(codeReq$, graphReq$)

  const resp$: Stream<Response> = sources.HTTP
    .select(INSERT_CODE_CELL)
    .thru(flatCatch)

  const reaction$: Stream<InsertCellResult> = resp$
    .flatMap(resp => {
      if (isOK(resp)) {
        const body = resp.body as gql.InsertCodeCellResponse

        const data = fp.get("data.insertCell", body)

        const action: InsertCellResult = {
          type: INSERT_CODE_CELL,
          status: AsyncStatus.Success,
          data: {
            sheetId: data.sheet.uuid,
            cells: data.sheet.cells.map((it:any) => it.uuid),
            cell: {
              worksheet: data.sheet.uuid,
              uuid: data.cell.uuid,
              script: data.cell.script,
              lang: "kotlin",
              spec: data.cell.spec
            }
          }


        }

        return most.of(action)
      }
      return most.empty()
    })

  return {
    HTTP: request$,
    ACTION: reaction$
  }
}

function reorderWorksheetCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const request$: Stream<RequestInput> = action$
    .filter(ofType(REORDER_WORKSHEET))
    .map(act => act as ReorderWorksheetAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({sheetId, cells}) => gql.reorderWorksheet(sheetId, cells))
    .map(body => ({
      url: '/graphql',
      category: REORDER_WORKSHEET,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !fp.get("body.errors", resp)
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(REORDER_WORKSHEET)
    .thru(flatCatch)

  const reaction$: Stream<ReorderWorksheetAction> = resp$
    .flatMap(resp => {
      // TODO map errors
      if (!isOK(resp))
        return most.empty()

      const body = resp.body as gql.ReorderWorksheetResponse
      const sheet = body.data.reorderWorksheet

      // TODO handle missing worksheet
      const cells = sheet.cells.map(it => it.uuid)

      return most.of(reorderWorksheet(sheet.uuid, cells, AsyncStatus.Success))
    })

  return {
    HTTP: request$,
    ACTION: reaction$,
  }
}

export default combineCycles(
  loadWorksheetCycle,
  afterCommitCycle,
  deleteCellCycle,
  insertCellCycle,
  reorderWorksheetCycle,
  )