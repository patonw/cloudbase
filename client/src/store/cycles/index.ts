import { Action } from 'redux';
import * as fp from 'lodash/fp'

import { Stream } from 'most'
import * as most from 'most'

import { combineCycles } from 'redux-cycles';
import { RequestInput } from '@cycle/http';

import * as gql from '../graphql'
import { AppState } from '../reducers';
import {
  AsyncAction, AsyncStatus,
  START_APP, loadToc,
  LOAD_WORKSHEET, LoadWorksheetAction,
  EXECUTE_CELL, ExecuteCellAction, afterCommit,
  DELETE_CELL, DeleteCellAction, deleteCell,
  INSERT_CODE_CELL, InsertCodeCellAction, InsertCellResult,
  INSERT_GRAPH_CELL, InsertGraphCellAction,
  REORDER_WORKSHEET, ReorderWorksheetAction, reorderWorksheet,
  CREATE_WORKSHEET, CreateWorksheetAction, createWorksheet,
  CREATE_PROCESS, createProcess, CreateProcessAction
} from '../actions';

function ofType(tag: any) {
  return (action: Action) => action.type === tag
}

function ofStatus(tag: AsyncStatus) {
  return (action: AsyncAction) => action.status === tag
}

interface ResponseSuccess {
  body: any
  request: any
}

interface ResponseError {
  response: ResponseSuccess
}

type Response = ResponseSuccess | ResponseError
const isOK = (resp: Response): resp is ResponseSuccess => !fp.isError(resp)

function flatCatch(resp$$: Stream<Stream<Response>>): Stream<Response> {
  return resp$$
    .flatMap((resp$) =>
      resp$
        .recoverWith((err: ResponseError) => most.of(err))
    )
}

function loggerCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION
  const dummy$ = action$
    .tap(act => { console.debug("ACTION", act) })
    .flatMap(act => most.empty())
  return {
    ACTION: dummy$
  }
}

function processCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION
  const state$: Stream<AppState> = sources.STATE

  const wsReq$: Stream<LoadWorksheetAction> = action$
    .filter(ofType(LOAD_WORKSHEET))
    .map(action => action as LoadWorksheetAction)
    .filter(ofStatus(AsyncStatus.Pending))

  const fromLoadSheet$ = wsReq$
    .sample((action, state) => ({ action, state }),
      wsReq$,
      state$)
    .filter(({ action, state }) => action.uuid! in state.worksheets)
    .map(({ action }) => createProcess(action.uuid))

  const request$: Stream<RequestInput> = action$
    .filter(ofType(CREATE_PROCESS))
    .map(it => it as CreateProcessAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map((act) => gql.createProcess(act.data.sheetId))
    .map(body => ({
      url: '/graphql',
      category: CREATE_PROCESS,
      method: "POST",
      send: body,
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(CREATE_PROCESS)
    .thru(flatCatch)


  const reaction$ = resp$.flatMap(resp => {
    // TODO handle errors
    if (!isOK(resp))
      return most.empty()

    const body = resp.body as gql.CreateProcessResponse
    const sheetId = body.data.createProcess.sheet.uuid
    const procId = body.data.createProcess.uuid

    return most.of(createProcess(sheetId, procId, AsyncStatus.Success))
  })

  return {
    ACTION: reaction$.merge(fromLoadSheet$),
    HTTP: request$,
  }
}

function loadWorksheetCycle(sources: any) {
  const request$: Stream<RequestInput> = sources.ACTION
    .filter(ofType(LOAD_WORKSHEET))
    .filter(ofStatus(AsyncStatus.Pending))
    .map((action: LoadWorksheetAction) => ({
      url: '/graphql',
      category: LOAD_WORKSHEET,
      method: "POST",
      send: gql.loadWorksheet(action.uuid),
      uuid: action.uuid,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.worksheet", resp)
    }))

  const result$ = (sources.HTTP)
    .select(LOAD_WORKSHEET)
    .flatMap((response$: Stream<Response>) =>
      response$
        .recoverWith((err: ResponseError) => most.of(err))
    )
    .map((resp: Response) => {
      let uuid = fp.get("body.data.worksheet.uuid", resp)

      if (isOK(resp) && !!uuid) {
        return {
          type: LOAD_WORKSHEET,
          status: AsyncStatus.Success,
          uuid,
          data: fp.get("body.data.worksheet", resp)
        }
      }

      uuid = uuid || fp.get("response.request.uuid", resp)
      let error = fp.get("body.errors", resp) || fp.get("response.body.errors", resp) || resp
      return {
        type: LOAD_WORKSHEET,
        status: AsyncStatus.Failure,
        uuid: uuid,
        error,
      }
    })

  return {
    HTTP: request$,
    ACTION: result$
  }
}

function afterCommitCycle(sources: any) {
  const state$: Stream<AppState> = sources.STATE
  const action$: Stream<Action> = sources.ACTION

  const dirtyCells$ = state$.map(state => state.dirty)
  const execSuccess$ = action$
    .filter(ofType(EXECUTE_CELL))
    .map(it => it as ExecuteCellAction)
    .filter(ofStatus(AsyncStatus.Success))

  const resolved$ = execSuccess$
    .sample(
      (action, dirty) => dirty[action.uuid],
      execSuccess$,
      dirtyCells$
    )
    .filter(fp.identity)
    .map(dirtyCell => afterCommit(dirtyCell))

  return {
    ACTION: resolved$
  }
}

function deleteCellCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const request$: Stream<RequestInput> = action$
    .filter(ofType(DELETE_CELL))
    .map(act => act as DeleteCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({ sheetId, cellId }) => gql.deleteCell(sheetId, cellId))
    .map(body => ({
      url: '/graphql',
      category: DELETE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.deleteCell.cell.uuid", resp)
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(DELETE_CELL)
    .thru(flatCatch)

  const reaction$ = resp$
    .flatMap(resp => {
      const cellId: string | undefined = fp.get("body.data.deleteCell.cell.uuid", resp)
      const sheetId: string | undefined = fp.get("body.data.deleteCell.sheet.uuid", resp)

      // TODO handle errors

      if (cellId && sheetId)
        return most.of(deleteCell(sheetId, cellId, AsyncStatus.Success))

      return most.empty()
    })

  return {
    HTTP: request$,
    ACTION: reaction$,
  }
}

function insertCellCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const codeReq$: Stream<RequestInput> = action$
    .filter(ofType(INSERT_CODE_CELL))
    .map(act => act as InsertCodeCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({ sheetId, index }) => gql.insertCodeCell(sheetId, index))
    .map(body => ({
      url: '/graphql',
      category: INSERT_CODE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.insertCell.cell.uuid", resp)
    }))

  const graphReq$: Stream<RequestInput> = action$
    .filter(ofType(INSERT_GRAPH_CELL))
    .map(act => act as InsertGraphCellAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({ sheetId, index }) => gql.insertGraphCell(sheetId, index))
    .map(body => ({
      url: '/graphql',
      category: INSERT_CODE_CELL,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !!fp.get("body.data.insertCell.cell.uuid", resp)
    }))

  const request$ = most.merge(codeReq$, graphReq$)

  const resp$: Stream<Response> = sources.HTTP
    .select(INSERT_CODE_CELL)
    .thru(flatCatch)

  const reaction$: Stream<InsertCellResult> = resp$
    .flatMap(resp => {
      if (isOK(resp)) {
        const body = resp.body as gql.InsertCodeCellResponse

        const data = fp.get("data.insertCell", body)

        const action: InsertCellResult = {
          type: INSERT_CODE_CELL,
          status: AsyncStatus.Success,
          data: {
            sheetId: data.sheet.uuid,
            cells: data.sheet.cells.map((it: any) => it.uuid),
            cell: {
              worksheet: data.sheet.uuid,
              uuid: data.cell.uuid,
              script: data.cell.script,
              lang: "kotlin",
              spec: data.cell.spec
            }
          }
        }

        return most.of(action)
      }

      return most.empty()
    })

  return {
    HTTP: request$,
    ACTION: reaction$
  }
}

function reorderWorksheetCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const request$: Stream<RequestInput> = action$
    .filter(ofType(REORDER_WORKSHEET))
    .map(act => act as ReorderWorksheetAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(({ sheetId, cells }) => gql.reorderWorksheet(sheetId, cells))
    .map(body => ({
      url: '/graphql',
      category: REORDER_WORKSHEET,
      method: "POST",
      send: body,
      ok: (resp: any) => resp.ok && !fp.get("body.errors", resp)
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(REORDER_WORKSHEET)
    .thru(flatCatch)

  const reaction$: Stream<ReorderWorksheetAction> = resp$
    .flatMap(resp => {
      // TODO map errors
      if (!isOK(resp))
        return most.empty()

      const body = resp.body as gql.ReorderWorksheetResponse
      const sheet = body.data.reorderWorksheet

      // TODO handle missing worksheet
      const cells = sheet.cells.map(it => it.uuid)

      return most.of(reorderWorksheet(sheet.uuid, cells, AsyncStatus.Success))
    })

  return {
    HTTP: request$,
    ACTION: reaction$,
  }
}

export function createWorksheetCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const request$: Stream<RequestInput> = action$
    .filter(ofType(CREATE_WORKSHEET))
    .map(act => act as CreateWorksheetAction)
    .filter(ofStatus(AsyncStatus.Pending))
    .map(act => act.data)
    .map(data => gql.createWorksheet(data.bookId, data.name))
    .map(body => ({
      url: '/graphql',
      category: CREATE_WORKSHEET,
      method: "POST",
      send: body,
    }))

  const resp$: Stream<Response> = sources.HTTP
    .select(CREATE_WORKSHEET)
    .thru(flatCatch)

  const reaction$: Stream<InsertCellResult> = resp$
    .flatMap(resp => {
      // TODO map errors
      if (!isOK(resp))
        return most.empty()

      const body = resp.body as gql.CreateWorksheetResponse
      const sheet = body.data.createWorksheet

      const bookId = fp.get("request.send.variables.bookId", resp)

      const action: CreateWorksheetAction = createWorksheet(bookId, sheet.name, AsyncStatus.Success)
      const tocAction = loadToc()

      return most.from([action, tocAction])
    })

  return {
    HTTP: request$,
    ACTION: reaction$,
  }
}

function startAppCycle(sources: any) {
  const action$: Stream<Action> = sources.ACTION

  const reaction$ = action$
    .filter(ofType(START_APP))
    .map(act => loadToc())

  return {
    ACTION: reaction$
  }
}

export default combineCycles(
  startAppCycle,
  loggerCycle,
  loadWorksheetCycle,
  afterCommitCycle,
  deleteCellCycle,
  insertCellCycle,
  reorderWorksheetCycle,
  createWorksheetCycle,
  processCycle,
)