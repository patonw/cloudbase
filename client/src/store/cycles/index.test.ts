import {loadWorksheetCycle } from './index'
import cycles from './index'

import * as most from 'most'
import { Stream } from 'most'

import { Action } from 'redux';

import { AppState } from '../reducers';
import * as act from '../actions'

const sourceEmpty = {
  ACTION: most.empty(),
  STATE: most.empty(),
  HTTP: {
    select: jest.fn((tag: string) => most.empty())
  },
}

async function dumpStream(stream: Stream<any>) {
  const init: any[] = []
  return await most.reduce((xs, x) => [...xs, x], init, stream)
}

describe('processCycle', () =>{
  it('creates a new process when none exists for a worksheet being loaded', () => {
    const state$: Stream<AppState> = most.empty()
    const action$: Stream<Action> = most.empty()

    // TODO
  })
})

describe('loadWorksheetCycle', () =>{
  it('issues a worksheet request from a LOAD_WORKSHEET action', () => {
    const action$: Stream<Action> = most.of(act.loadWorksheet('theWorksheetId'))
    const sources = {
      ...sourceEmpty,
      ACTION: action$,
    }
    const result = loadWorksheetCycle(sources)
    most.observe((req) => expect(req).toMatchSnapshot(), result.HTTP)
  })

  it('processes GraphQL response', async () => {
    const httpSelector = jest.fn((selector) => most.of(most.of({
      body: {
        data: {
          worksheet: {
            uuid: "theWorksheetId"
          }
        }
      }
    })))

    const sources = {
      ...sourceEmpty,
      HTTP: {
        select: httpSelector,
      }
    }

    const result = loadWorksheetCycle(sources)
    expect(httpSelector.mock.calls).toMatchSnapshot()

    const actions = await dumpStream(result.ACTION)
    expect(actions).toMatchSnapshot()
  })

  it('handles errors by dispatching async failure', async() => {
    const httpSelector = jest.fn((selector) => most.of(most.throwError({
      response: {
        request: {
          uuid: "badUUID"
        }
      }

    } as any)))

    const sources = {
      ...sourceEmpty,
      HTTP: {
        select: httpSelector,
      }
    }

    const result = loadWorksheetCycle(sources)
    expect(httpSelector.mock.calls).toMatchSnapshot()

    const actions = await dumpStream(result.ACTION)
    expect(actions).toMatchSnapshot()
  })
})