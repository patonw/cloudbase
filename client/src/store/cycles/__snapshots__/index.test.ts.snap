// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadWorksheetCycle handles errors by dispatching async failure 1`] = `
Array [
  Array [
    "LOAD_WORKSHEET",
  ],
]
`;

exports[`loadWorksheetCycle handles errors by dispatching async failure 2`] = `
Array [
  Object {
    "error": Object {
      "response": Object {
        "request": Object {
          "uuid": "badUUID",
        },
      },
    },
    "status": 3,
    "type": "LOAD_WORKSHEET",
    "uuid": "badUUID",
  },
]
`;

exports[`loadWorksheetCycle issues a worksheet request from a LOAD_WORKSHEET action 1`] = `
Object {
  "category": "LOAD_WORKSHEET",
  "method": "POST",
  "ok": [Function],
  "send": Object {
    "query": "
  query loadSheet($sheetId: ID!) {
    worksheet(sheetId: $sheetId) {
      uuid
      name
      cells {
        __typename
        uuid
        script
        ... on GraphCell {
          spec
        }
      }
    }
  }",
    "variables": Object {
      "sheetId": "theWorksheetId",
    },
  },
  "url": "/graphql",
  "uuid": "theWorksheetId",
}
`;

exports[`loadWorksheetCycle processes GraphQL response 1`] = `
Array [
  Array [
    "LOAD_WORKSHEET",
  ],
]
`;

exports[`loadWorksheetCycle processes GraphQL response 2`] = `
Array [
  Object {
    "data": Object {
      "uuid": "theWorksheetId",
    },
    "status": 2,
    "type": "LOAD_WORKSHEET",
    "uuid": "theWorksheetId",
  },
]
`;
